Update EIP-7873: Handling of EOF containers in existing TX types. (#9669)
ethereum/EIPs: EIPS/eip-7873.md

Under transaction validation rules `initcodes` are not validated for conforming to the EOF specification. They are only validated when accessed via `TXCREATE`. This avoids potential DoS attacks of the mempool. If during the execution of an `InitcodeTransaction` no `TXCREATE` instruction is called, such transaction is still valid.

> Other creation transactions that support contract creation (specifically type 0 "Frontier," type 1 "AccessList," type 2 "FeeMarket" transactions with an empty `to` field) will not attempt to parse EOF containers in their `input` field and will execute the code as non-EOF code. This will result in immediately executing the `INVALID`(`0xEF`) instruction (see [EIP-141](./eip-141.md)) and halting.

Reviewer: ```suggestion
Other creation transactions that support contract creation (specifically type 0 "Frontier," type 1 "AccessList," type 2 "FeeMarket" transactions with an empty `to` field) will not attempt to parse EOF containers in their `input` field and will execute the code as non-EOF code. This will result in immediately executing the undefined `0xEF` instruction and halting.
```

Three existing transaction types (type 0 "Frontier," type 1 "AccessList," type 2 "FeeMarket") accept code as part of their input data in certain configurations. This code can start with `EF` as it is initcode and not a deployed contract.

One possible way of handling potential EOF containers in these transactions is to make them invalid if they are attempting to execute an EOF container as initcode. Specifically the transactions is invalid if the input data begins with `0xef00` and if the `to` field is set to nil, signaling a contract creation transaction. This would make blocks containing these transactions invalid, whereas before the fork including this EIP they would have been valid.

One impact that would be seen by this would be block builders, as they would need to ensure that the transactions that are invalid are not included in a block. There is precedent for this, as [EIP-7623](./eip-7623.md) established a new floor for the gas limit in transactions, prior to its adoption the limit was lower.  However, block builders have an existing check on gas limits and transactions and that change was adjusting the formula and constants.

> We should also consider the treatment [EIP-7702](./eip-7702) delegation designations receive if they show up as code in a contract creation transaciton.  The transaction is valid, is executed as plain bytecode, ad the first operation executed is `0xEF`, which as per [EIP-141](eip-141) is the designated invalid opcode and would cause the whole transction to abort.

Reviewer: EF is not the designated invalid opcode, FE is.

> Letting the inputdata of a type 0, 1, and 2 contract creation transaction execue as plain EVM code irrespective of the magic `0xEF00` byes keeps a consistent behavior in client code and keeps block builders from having to update their logic in older transaction types.

Reviewer: ```suggestion
Letting the inputdata of a type 0, 1, and 2 contract creation transaction execue as plain EVM code irrespective of the magic `0xEF00` bytes keeps a consistent behavior in client code and keeps block builders from having to update their logic in older transaction types.
```

The transactions of the new type are invalid until this change activates.

> Contract creation options do not change for legacy bytecode, including how existing transactions behave when encountering code that may look like an EOF container.

Reviewer 2: ```suggestion
Contract creation options do not change for legacy bytecode, including how existing transactions with `to: nil` behave when encountering code in their calldata that may look like an EOF container.
```

Otherwise
> encountering code that may look like an EOF container.

is ambiguous.
